src/libutil/url.cc

```diff
@@ -88,17 +88,22 @@ std::map<std::string, std::string> decodeQuery(const std::string & query)
     return result;
 }
 
-std::string percentEncode(std::string_view s)
+const static std::string allowedInQuery = ":@/?";
+const static std::string allowedInPath = ":@/";
+
+std::string percentEncode(std::string_view s, std::string_view keep)
 {
     std::string res;
     for (auto & c : s)
+        // unreserved + keep
         if ((c >= 'a' && c <= 'z')
             || (c >= 'A' && c <= 'Z')
             || (c >= '0' && c <= '9')
-            || strchr("-._~!$&'()*+,;=:@", c))
+            || strchr("-._~", c)
+            || keep.find(c) != std::string::npos)
```

Comment:
Why do we need to make this stricter? Was the old behaviour incorrect?

Reply:
The removed characters belong to a "reserved" set, which have a function in some parts of the URL (such as `&` being a query param separator).

It is always safe to escape characters even if they are allowed, but `keep` has been added for when you know which characters are allowed in the part of the URL you are encoding.